testutils/src/index.ts

```diff
@@ -74,6 +74,53 @@ export function testEmission<T, U, V>(
   return done.promise;
 }
 
+/**
+ * Convert a signal into an array of promises.
+ *
+ * @param signal - The signal we are listening to.
+ * @param numberValues - The number of values to store.
+ *
+ * @returns a Promise that resolves with an array of `(sender, args)` pairs.
+ */
+export function signalToPromises<T, U>(
+  signal: ISignal<T, U>,
+  numberValues: number
+): Promise<[T, U]>[] {
+  const values: Promise<[T, U]>[] = new Array(numberValues);
+  const resolvers: Array<((value: [T, U]) => void)> = new Array(numberValues);
+
+  for (let i = 0; i < numberValues; i++) {
+    values[i] = new Promise<[T, U]>(resolve => {
+      resolvers[i] = resolve;
+    });
+  }
+
+  let current = 0;
+  function slot(sender: T, args: U) {
+    resolvers[current++]([sender, args]);
+    if (current === numberValues) {
+      cleanup();
+    }
+  }
+  signal.connect(slot);
+
+  function cleanup() {
+    signal.disconnect(slot);
+  }
+  return values;
```

Comment:
If we don't care about cleanup (and aren't returning that function), then we  can remove  the `function cleanup` declaration

Reply:
We use it internally when we're done collecting signals.

Reply:
Collect emissions, that is.

Reply:
Nevermind, I see you added cleanup to the slot function.

I was thinking of having a go at implementing this to return an interable (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators#User-defined_iterables) so that the user doesn't have to pass in the number of values, but looks like the version as-is is working OK.

Reply:
We have been using `Phosphor` iterators in general.  I think changing it to `signalToIterator(): IIterator<Promise<T,U>>` makes sense.  `signalToPromise` would just call `.next()` on that iterator and return it.

Reply:
Although we wouldn't be able to remove the slot, because it would have to be an inifinite stream that hands out promises (unless we offer an optional bound).

Reply:
> Although we wouldn't be able to remove the slot, because it would have to be an inifinite stream that hands out promises (unless we offer an optional bound).

Yeah cleanup would require returning another callback.

> We have been using Phosphor iterators in general.

Do they implement the `@@iterator` method so that can use them in destructuring and for...of loops? I personally like being able to use those language features.

Reply:
Because of the requirement of the native iterator to allocate per iteration, there isn't a way to reconcile phosphor iterators with the native one. And because they are more efficient and already endemic to the codebase, it's best to stay consistent.

Reply:
@saulshanabrook, we have docs about the `Phosphor` iterators [here](https://github.com/jupyterlab/jupyterlab/blob/master/docs/source/developer/patterns.rst#data-structures).  PRs welcome to improve those docs!